In [3]:
import numpy as np
from rref import *

## Example 6.21

In [15]:
P = np.array([[0.5, 0.4, 0.6],
              [0.2, 0.2, 0.3],
              [0.3, 0.4, 0.1]])
P

array([[0.5, 0.4, 0.6],
       [0.2, 0.2, 0.3],
       [0.3, 0.4, 0.1]])

In [16]:
x0 = np.array([0,1,0]) # intitial state vector (lion starts in reserve 2)
print(f"x0 = {x0}")

x1 = P@x0
print(f"x1 = {x1}")

x2 = P@x1
print(f"x2 = {x2}")
x3 = P@x2
print(f"x3 = {x3}")
x4 = P@x3
print(f"x4 = {x4}")
x5 = P@x4
print(f"x5 = {x5}")
x6 = P@x5
print(f"x6 = {x6}")


x0 = [0 1 0]
x1 = [0.4 0.2 0.4]
x2 = [0.52 0.24 0.24]
x3 = [0.5   0.224 0.276]
x4 = [0.5052 0.2276 0.2672]
x5 = [0.50396 0.22672 0.26932]
x6 = [0.50426  0.226932 0.268808]


Can use a for loop to make it more robust:

In [17]:
x = np.array([0,1,0]) # initial state vector (lion starts in reserve 2)
print(f"x0 = {x}")

for i in range(1,16):
    x = P@x
    print(f"x{i} = {x}")

x0 = [0 1 0]
x1 = [0.4 0.2 0.4]
x2 = [0.52 0.24 0.24]
x3 = [0.5   0.224 0.276]
x4 = [0.5052 0.2276 0.2672]
x5 = [0.50396 0.22672 0.26932]
x6 = [0.50426  0.226932 0.268808]
x7 = [0.5041876 0.2268808 0.2689316]
x8 = [0.50420508 0.22689316 0.26890176]
x9 = [0.50420086 0.22689018 0.26890896]
x10 = [0.50420188 0.2268909  0.26890722]
x11 = [0.50420163 0.22689072 0.26890764]
x12 = [0.50420169 0.22689076 0.26890754]
x13 = [0.50420168 0.22689075 0.26890757]
x14 = [0.50420168 0.22689076 0.26890756]
x15 = [0.50420168 0.22689076 0.26890756]


The vector $[0.50420168 0.22689076 0.26890756] = [50\%,23\%,27\%]$ is the steady state solution.

## Example 6.24

The identity matrix can be created using the `np.eye` command.

In [19]:
# We want to find the eigenvector corresponding to eigenvalue 1.
# This is the steady state solution.
rref(P-np.eye(3))

(array([[ 1.     ,  0.     , -1.875  ],
        [ 0.     ,  1.     , -0.84375],
        [ 0.     ,  0.     ,  0.     ]]),
 (0, 1))

In [20]:
v = np.array([1.875,0.84375,1]) # reading off the solution from the rref above
q = v/sum(v) # normalize it to a probability vector: sum of components must be 1
print(q) # display steady state solution


[0.50420168 0.22689076 0.26890756]


Note: this is the same vector we computed in Example 6.21 by using the for-loop and seeing what the $x_k$ converged to.